crates/sui-framework/packages/sui-framework/sources/coin_metadata_registry.move

```diff
@@ -0,0 +1,208 @@
+module sui::coin_metadata_registry;
+
+use std::string::String;
+use sui::balance::Supply;
+use sui::coin::{TreasuryCap, CoinMetadata, RegulatedCoinMetadata};
+use sui::dynamic_field;
+use sui::transfer::Receiving;
+
+const ENotSystemAddress: u64 = 0;
+const EMetadataAlreadyExists: u64 = 1;
+const EMetadataNotFound: u64 = 2;
+const EAlreadyClaimed: u64 = 3;
+const EMetadataCapNotClaimed: u64 = 4;
+const EAlreadyRegulated: u64 = 5;
+
+public struct Registry has key { id: UID }
+
+// === Coin Metadata ===
+public struct CoinMetadataKey<phantom T>() has copy, drop, store;
+
+public struct CoinMetadataCap<phantom T> has key, store { id: UID }
+
+public struct MetadataCapClaimed<phantom T>() has copy, drop, store;
+
+public struct Metadata<phantom T> has store {
+    decimals: u8,
+    name: String,
+    symbol: String,
+    description: String,
+    icon_url: String,
+    is_regulated: Option<bool>,
+    supply: Option<Supply<T>>,
+}
+
+public fun metadata<T>(registry: &Registry): &Metadata<T> {
+    assert!(registry.exists<T>(), EMetadataNotFound);
+    dynamic_field::borrow(
+        &registry.id,
+        CoinMetadataKey<T>(),
+    )
+}
+
+public fun mark_regulated<T>(registry: &mut Registry, _: &RegulatedCoinMetadata<T>) {
+    assert!(registry.exists<T>(), EMetadataNotFound);
+    let metadata: &mut Metadata<T> = dynamic_field::borrow_mut(
+        &mut registry.id,
+        CoinMetadataKey<T>(),
+    );
+
+    assert!(metadata.is_regulated.is_none(), EAlreadyRegulated);
+    metadata.is_regulated.fill(true);
+}
+
+/// Allows the owner to freeze the currency by destroying the treasury cap.
+public fun freeze_supply<T>(registry: &mut Registry, mut cap: TreasuryCap<T>) {
+    assert!(registry.exists<T>(), EMetadataNotFound);
+    assert!(
+        dynamic_field::exists_(cap.treasury_cap_uid_mut(), MetadataCapClaimed<T>()),
+        EMetadataCapNotClaimed,
+    );
+
+    let metadata: &mut Metadata<T> = dynamic_field::borrow_mut(
+        &mut registry.id,
+        CoinMetadataKey<T>(),
+    );
+
+    metadata.supply.fill(cap.destroy_treasury_cap())
+}
+
+public fun migrate<T>(registry: &mut Registry, metadata_v1: CoinMetadata<T>) {
+    registry.attach_metadata(from_v1(&metadata_v1));
+    metadata_v1.destroy_metadata()
+}
+
+public fun migrate_frozen<T>(registry: &mut Registry, metadata_v1: &CoinMetadata<T>) {
```

Comment:
Any metadata can be registered by reference, and it might create conflicts with (just) `migrate` - we must account for that. Simple test:

```
let metadata = ...
migrate_frozen(&metadata);
migrate(metadata); // we want this to work to delete the original object
```